File: results/2026-06-26_103905/tasks/12-pr-label-assigner/typescript-bun-opus48-1m-xhigh/generated-code/tests/act-harness.ts
#!/usr/bin/env bun
/**
 * Act-based integration harness for the PR Label Assigner workflow.
 *
 * Per the task spec, EVERY integration test case is executed through the real
 * GitHub Actions workflow via `act` (nektos/act) rather than by calling the
 * script directly. For each case this harness:
 *
 *   1. Builds a throwaway git repo containing the project + that case's
 *      changed-files fixture (written to examples/changed-files.json, the path
 *      the workflow reads by default).
 *   2. Runs `act push --rm` against the workflow inside that repo.
 *   3. Appends the full act output to ./act-result.txt, clearly delimited.
 *   4. Asserts: act exited 0, both jobs report "Job succeeded", the workflow
 *      emitted the EXACT expected label set (via the __PR_LABELS__= marker the
 *      CLI prints), and the GITHUB_OUTPUT-derived label count matches.
 *
 * Run with:  bun run tests/act-harness.ts
 * Exits non-zero if any case fails.
 */
import {
  cpSync,
  mkdtempSync,
  writeFileSync,
  appendFileSync,
  rmSync,
  existsSync,
} from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

const PROJECT_ROOT = join(import.meta.dir, "..");
const ACT_RESULT = join(PROJECT_ROOT, "act-result.txt");
const WORKFLOW_REL = ".github/workflows/pr-label-assigner.yml";
const ACT_IMAGE = "act-ubuntu-pwsh:latest";

// Files/dirs copied into each throwaway repo. node_modules is intentionally
// omitted — the script has zero runtime dependencies and the workflow installs
// Bun itself.
const COPY_ITEMS = [
  "src",
  "tests",
  "examples",
  ".github",
  "package.json",
  "tsconfig.json",
  ".actrc",
];

interface TestCase {
  /** Short, filesystem-safe identifier. */
  name: string;
  /** Human description of what the case exercises. */
  description: string;
  /** The mock PR's changed file paths. */
  changedFiles: string[];
  /** EXACT expected machine-readable label list (priority order, comma-sep). */
  expectedLabels: string;
  /** EXACT expected label count (round-tripped through GITHUB_OUTPUT). */
  expectedCount: number;
}

// The cases are evaluated against the default examples/labeler.config.json:
//   documentation: docs/**, *.md         (priority 1)
//   api:           src/api/**            (priority 10)
//   tests:         *.test.*, **/*.spec.ts (priority 5)
//   ci:            .github/**            (priority 8)
//   frontend:      src/web/**, *.css, *.html (priority 3)
const CASES: TestCase[] = [
  {
    name: "multi-label-priority",
    description:
      "A file can yield multiple labels; the union is emitted in priority order.",
    changedFiles: [
      "docs/getting-started.md", // documentation (p1)
      "src/api/users.ts", // api (p10)
      "src/api/users.test.ts", // api (p10) + tests (p5)
      "src/web/app.tsx", // frontend (p3)
    ],
    // priority desc: api(10), tests(5), frontend(3), documentation(1)
    expectedLabels: "api,tests,frontend,documentation",
    expectedCount: 4,
  },
  {
    name: "docs-and-ci",
    description:
      "Directory and basename globs both fire; ci(8) outranks documentation(1).",
    changedFiles: [
      "docs/intro.md", // documentation (docs/**)
      "README.md", // documentation (*.md) — deduped, no double label
      ".github/workflows/deploy.yml", // ci (.github/**)
    ],
    expectedLabels: "ci,documentation",
    expectedCount: 2,
  },
  {
    name: "no-matches",
    description: "When nothing matches, the label set is empty (not an error).",
    changedFiles: ["LICENSE", "Makefile", "src/core/engine.ts"],
    expectedLabels: "",
    expectedCount: 0,
  },
];

/** Strip ANSI escape sequences so we can match plain text in act's output. */
function stripAnsi(s: string): string {
  // eslint-disable-next-line no-control-regex
  return s.replace(/\x1b\[[0-9;]*[A-Za-z]/g, "");
}

/** Run a command synchronously, throwing with context on failure. */
function run(cmd: string[], cwd: string): void {
  const proc = Bun.spawnSync(cmd, { cwd, stdout: "pipe", stderr: "pipe" });
  if (proc.exitCode !== 0) {
    throw new Error(
      `command failed (${proc.exitCode}): ${cmd.join(" ")}\n` +
        proc.stderr.toString(),
    );
  }
}

/** Create a throwaway git repo for one test case and return its path. */
function setupRepo(testCase: TestCase): string {
  const dir = mkdtempSync(join(tmpdir(), `act-pr-label-${testCase.name}-`));

  for (const item of COPY_ITEMS) {
    const src = join(PROJECT_ROOT, item);
    if (existsSync(src)) cpSync(src, join(dir, item), { recursive: true });
  }

  // Inject this case's changed-files fixture into the path the workflow reads.
  writeFileSync(
    join(dir, "examples", "changed-files.json"),
    JSON.stringify(testCase.changedFiles, null, 2) + "\n",
  );

  // act derives event context from git; use a branch the workflow listens on.
  run(["git", "init", "-q", "-b", "main"], dir);
  run(["git", "config", "user.email", "ci@example.com"], dir);
  run(["git", "config", "user.name", "CI"], dir);
  run(["git", "add", "-A"], dir);
  run(["git", "commit", "-q", "-m", `fixture: ${testCase.name}`], dir);

  return dir;
}

/** Run the workflow under act for the given repo; return exit code + output. */
function runAct(dir: string): { exitCode: number; output: string } {
  const proc = Bun.spawnSync(
    [
      "act",
      "push",
      "--rm",
      "--pull=false", // the custom image is local-only; never attempt a registry pull
      "-P",
      `ubuntu-latest=${ACT_IMAGE}`,
      "-W",
      WORKFLOW_REL,
    ],
    { cwd: dir, stdout: "pipe", stderr: "pipe" },
  );
  const output = stripAnsi(proc.stdout.toString() + "\n" + proc.stderr.toString());
  return { exitCode: proc.exitCode ?? -1, output };
}

/**
 * Return the text after `marker` on the first line that contains it (with
 * surrounding whitespace trimmed), or null if no line contains the marker.
 */
function valueAfter(output: string, marker: string): string | null {
  for (const line of output.split(/\r?\n/)) {
    const idx = line.indexOf(marker);
    if (idx !== -1) return line.slice(idx + marker.length).trim();
  }
  return null;
}

interface CaseResult {
  name: string;
  passed: boolean;
  failures: string[];
}

/** Evaluate all assertions for one completed act run. */
function assertCase(
  testCase: TestCase,
  exitCode: number,
  output: string,
): CaseResult {
  const failures: string[] = [];

  if (exitCode !== 0) {
    failures.push(`act exited with code ${exitCode} (expected 0)`);
  }

  // EXACT expected label set, via the CLI's machine-readable marker line.
  const labels = valueAfter(output, "__PR_LABELS__=");
  if (labels === null) {
    failures.push("no __PR_LABELS__= marker found in act output");
  } else if (labels !== testCase.expectedLabels) {
    failures.push(
      `labels mismatch: got "${labels}", expected "${testCase.expectedLabels}"`,
    );
  }

  // EXACT label count, round-tripped through GITHUB_OUTPUT -> step output.
  const count = valueAfter(output, "Label count:");
  if (count === null) {
    failures.push("no 'Label count:' line found (GITHUB_OUTPUT round-trip)");
  } else if (count !== String(testCase.expectedCount)) {
    failures.push(
      `count mismatch: got "${count}", expected "${testCase.expectedCount}"`,
    );
  }

  // Both jobs (test + assign-labels) must report success.
  const succeeded = (output.match(/Job succeeded/g) ?? []).length;
  if (succeeded < 2) {
    failures.push(
      `expected 2 "Job succeeded" (one per job), found ${succeeded}`,
    );
  }
  if (/Job failed/.test(output)) {
    failures.push('output contains "Job failed"');
  }

  return { name: testCase.name, passed: failures.length === 0, failures };
}

/** Append a clearly delimited record of one case to act-result.txt. */
function recordResult(
  testCase: TestCase,
  exitCode: number,
  output: string,
  result: CaseResult,
): void {
  const banner = "=".repeat(78);
  const verdict = result.passed ? "PASS" : "FAIL";
  let block = `\n${banner}\n`;
  block += `TEST CASE: ${testCase.name}  [${verdict}]\n`;
  block += `DESCRIPTION: ${testCase.description}\n`;
  block += `CHANGED FILES: ${JSON.stringify(testCase.changedFiles)}\n`;
  block += `EXPECTED LABELS: "${testCase.expectedLabels}" (count ${testCase.expectedCount})\n`;
  block += `ACT EXIT CODE: ${exitCode}\n`;
  if (result.failures.length > 0) {
    block += `FAILURES:\n  - ${result.failures.join("\n  - ")}\n`;
  }
  block += `${banner}\n`;
  block += `----- BEGIN ACT OUTPUT (${testCase.name}) -----\n`;
  block += output.trimEnd() + "\n";
  block += `----- END ACT OUTPUT (${testCase.name}) -----\n`;
  appendFileSync(ACT_RESULT, block);
}

async function main(): Promise<number> {
  // Start a fresh act-result.txt for this run.
  const header =
    `PR Label Assigner — act integration results\n` +
    `Workflow: ${WORKFLOW_REL}\n` +
    `Image: ${ACT_IMAGE}\n` +
    `Cases: ${CASES.length}\n`;
  writeFileSync(ACT_RESULT, header);

  const results: CaseResult[] = [];

  for (const testCase of CASES) {
    console.log(`\n>>> Running act for case: ${testCase.name}`);
    const dir = setupRepo(testCase);
    try {
      const { exitCode, output } = runAct(dir);
      const result = assertCase(testCase, exitCode, output);
      recordResult(testCase, exitCode, output, result);
      results.push(result);
      console.log(
        `    ${result.passed ? "PASS" : "FAIL"} — ${testCase.name}` +
          (result.passed ? "" : `\n      ${result.failures.join("\n      ")}`),
      );
    } finally {
      // Throwaway repo; the act output is already saved to act-result.txt.
      try {
        rmSync(dir, { recursive: true, force: true });
      } catch {
        /* best effort */
      }
    }
  }

  const passed = results.filter((r) => r.passed).length;
  const summary =
    `\n${"#".repeat(78)}\n` +
    `SUMMARY: ${passed}/${results.length} cases passed\n` +
    results
      .map((r) => `  ${r.passed ? "PASS" : "FAIL"}  ${r.name}`)
      .join("\n") +
    `\n${"#".repeat(78)}\n`;
  appendFileSync(ACT_RESULT, summary);
  console.log(summary);

  return passed === results.length ? 0 : 1;
}

main().then((code) => process.exit(code));

File: results/2026-07-01_184135/tasks/12-pr-label-assigner/default-fable5-high/generated-code/act_harness.py
"""End-to-end test harness: run every labeler test case through the
GitHub Actions workflow via `act` (nektos/act).

For each case it builds a temp git repo containing the project files plus
that case's fixture data (the mocked PR changed-file list and the rules
config), runs `act push --rm`, appends the output to act-result.txt, and
asserts:
  * act exited 0
  * the output contains the EXACT expected 'LABELS: ...' line
  * the output contains the exact 'FINAL LABEL SET: ...' echoed from the
    step output (proves data flowed through GITHUB_OUTPUT)
  * both workflow jobs report 'Job succeeded'
"""
import json
import os
import shutil
import subprocess
import sys
import tempfile

ROOT = os.path.dirname(os.path.abspath(__file__))
RESULT_FILE = os.path.join(ROOT, "act-result.txt")

# Files the temp repo needs to run the workflow. tests/test_workflow.py is
# host-only (needs actionlint), so it is deliberately not copied.
PROJECT_FILES = [
    "labeler.py",
    "tests/__init__.py",
    "tests/test_labeler.py",
    ".github/workflows/pr-label-assigner.yml",
    ".actrc",  # maps ubuntu-latest to the local runner image
]

DEFAULT_RULES = {
    "rules": [
        {"pattern": "docs/**", "labels": ["documentation"]},
        {"pattern": "src/api/**", "labels": ["api", "backend"]},
        {"pattern": "src/**", "labels": ["source"]},
        {"pattern": "*.test.*", "labels": ["tests"]},
        {"pattern": ".github/**", "labels": ["ci"]},
    ]
}

PRIORITY_RULES = {
    "rules": [
        {"pattern": "src/**", "labels": ["source"]},
        {"pattern": "src/generated/**", "labels": ["generated", "skip-review"], "priority": 5},
        {"pattern": "docs/**", "labels": ["documentation"]},
    ]
}

CASES = [
    {
        "name": "case1-mixed-changes",
        "rules": DEFAULT_RULES,
        "changed_files": [
            "docs/getting-started.md",
            "src/api/users.py",
            "src/core/engine.py",
            "webapp/app.test.js",
        ],
        # docs -> documentation; src/api -> api,backend (+source via src/**);
        # src/core -> source; *.test.* -> tests
        "expected": "api,backend,documentation,source,tests",
    },
    {
        "name": "case2-priority-conflict",
        "rules": PRIORITY_RULES,
        "changed_files": ["src/generated/schema_pb2.py", "src/main.py"],
        # schema_pb2.py matches src/** (prio 0) AND src/generated/** (prio 5):
        # the high-priority rule suppresses 'source' for that file. main.py
        # still gets 'source' — priority is resolved per file.
        "expected": "generated,skip-review,source",
    },
    {
        "name": "case3-no-matching-rules",
        "rules": DEFAULT_RULES,
        "changed_files": ["LICENSE", "Makefile"],
        "expected": "(none)",
    },
]

JOB_NAMES = ["unit-tests", "assign-labels"]


def build_repo(case, repo):
    """Assemble a temp git repo: project files + this case's fixtures."""
    for rel in PROJECT_FILES:
        dst = os.path.join(repo, rel)
        os.makedirs(os.path.dirname(dst) or repo, exist_ok=True)
        shutil.copy2(os.path.join(ROOT, rel), dst)
    os.makedirs(os.path.join(repo, "fixtures"), exist_ok=True)
    with open(os.path.join(repo, "fixtures", "rules.json"), "w") as fh:
        json.dump(case["rules"], fh, indent=2)
    with open(os.path.join(repo, "fixtures", "changed_files.txt"), "w") as fh:
        fh.write("\n".join(case["changed_files"]) + "\n")
    git_env = {**os.environ,
               "GIT_AUTHOR_NAME": "harness", "GIT_AUTHOR_EMAIL": "h@test",
               "GIT_COMMITTER_NAME": "harness", "GIT_COMMITTER_EMAIL": "h@test"}
    for cmd in (["git", "init", "-q", "-b", "main"],
                ["git", "add", "-A"],
                ["git", "commit", "-qm", f"fixture: {case['name']}"]):
        subprocess.run(cmd, cwd=repo, env=git_env, check=True,
                       capture_output=True)


def run_case(case, log):
    repo = tempfile.mkdtemp(prefix=f"labeler-{case['name']}-")
    failures = []
    try:
        build_repo(case, repo)
        proc = subprocess.run(
            ["act", "push", "--rm", "--pull=false"],
            cwd=repo, capture_output=True, text=True, timeout=420,
        )
        output = proc.stdout + proc.stderr
        log.write(f"\n{'=' * 72}\n=== TEST CASE: {case['name']}\n"
                  f"=== changed files: {case['changed_files']}\n"
                  f"=== expected label set: {case['expected']}\n"
                  f"=== act exit code: {proc.returncode}\n{'=' * 72}\n")
        log.write(output)

        if proc.returncode != 0:
            failures.append(f"act exited {proc.returncode}, expected 0")
        expected_line = f"LABELS: {case['expected']}"
        if expected_line not in output:
            failures.append(f"missing exact output {expected_line!r}")
        final_line = f"FINAL LABEL SET: {case['expected']}"
        if final_line not in output:
            failures.append(f"missing exact output {final_line!r}")
        for job in JOB_NAMES:
            if not any(job in line and "Job succeeded" in line
                       for line in output.splitlines()):
                failures.append(f"job {job!r} did not report 'Job succeeded'")
    finally:
        shutil.rmtree(repo, ignore_errors=True)
    return failures


def main():
    # Pre-flight on the host: actionlint + workflow structure tests + the
    # full TDD unit suite (the same unit suite also runs inside act).
    for desc, cmd in [
        ("actionlint", ["actionlint", ".github/workflows/pr-label-assigner.yml"]),
        ("workflow structure tests", [sys.executable, "-m", "unittest", "tests.test_workflow", "-v"]),
        ("unit test suite", [sys.executable, "-m", "unittest", "discover", "-s", "tests", "-t", "."]),
    ]:
        proc = subprocess.run(cmd, cwd=ROOT, capture_output=True, text=True)
        status = "PASS" if proc.returncode == 0 else "FAIL"
        print(f"[pre-flight] {desc}: {status}")
        if proc.returncode != 0:
            print(proc.stdout + proc.stderr)
            return 1

    all_failures = {}
    with open(RESULT_FILE, "w") as log:
        log.write("act end-to-end results for pr-label-assigner\n")
        for case in CASES:
            print(f"[act] running {case['name']} ...", flush=True)
            failures = run_case(case, log)
            verdict = "PASS" if not failures else "FAIL: " + "; ".join(failures)
            log.write(f"\n>>> CASE VERDICT [{case['name']}]: {verdict}\n")
            print(f"[act] {case['name']}: {verdict}")
            if failures:
                all_failures[case["name"]] = failures
        log.write(f"\n{'=' * 72}\nOVERALL: "
                  f"{'ALL CASES PASSED' if not all_failures else 'FAILURES: ' + repr(all_failures)}\n")

    if all_failures:
        print(f"FAILED: {all_failures}")
        return 1
    print(f"All {len(CASES)} act test cases passed. Results in {RESULT_FILE}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
